hmem_cuda: add gdrcopy as an alternative mechanism for cuda memory copy

wzamazon opened the pull request:
This patch added gdrcopy as an alternative mechanism for cuda memory
copy to cudaMemcpy. It is to address the issue:

    https://github.com/ofiwg/libfabric/issues/6124

which is using cudaMemcpy can cause deadlock when application lauches
kernel on the targeted buffer.

This patch include the following changes:

1. Two arguments --with-gdrcopy and --enable-gdrcopy-dlopen was
   added to the configure script. User are expected to used them
   together with the --with-cuda argument.

2. Four functions:
           cuda_gdrcopy_hmem_init(),
           cuda_gdrcopy_hmem_cleanup(),
           cuda_gdrcopy_from_dev(),
           cuda_gdrcopy_to_dev(),
   was implemented and was called by repective cuda_hmem functions

3. Two additional functions:
	   ofi_hmem_cuda_dev_register() and
           ofi_hmem_cuda_dev_dereigster()
   are added. To use gdrcopy, user need to call the first one during
   memory registration, and passed the resulted handle to
   ofi_copy_to_hmem_iov().

4. An environment variable FI_HMEM_CUDA_USE_GDRCOPY was introduced,
   It is set to 1 by default. User can use it to switch back to use
   cudaMemcpy.

Signed-off-by: Wei Zhang <wzam@amazon.com>

shefty commented on Makefile.am:
```diff
@@ -43,6 +43,7 @@ common_srcs =				\
 	src/hmem.c			\
 	src/hmem_rocr.c			\
 	src/hmem_cuda.c			\
+	src/hmem_cuda_gdrcopy.c		\
```
Can we just fold this into hmem_cuda.c and skip adding a new source file?

Also, please update the commit message to describe why we want gdrcopy instead of cuda memcpy.

wzamazon replied:
We can fold the file into hmem_cuda.c, but it would make the resulted file not easy to read IMO. I would prefer to introduce a new file.

Updated the commit message to add the reasoning behind this PR.

shefty commented on src/hmem_cuda.c:
```diff
@@ -102,6 +102,14 @@ CUresult ofi_cuInit(unsigned int flags)
+	err = cuda_gdrcopy_to_dev(device, dev, host, size);
+	if (!err)
+		return 0; 
+#endif
```
ifdefs in the middle of functions make it a pain to read the code.  It's preferable to always define cuda_gdrcopy_to_dev, and just make it a no-op or always fail if gdrcopy is not enabled.  The compiler should toss out the resulting code.

Separate, why does this fall through to cuda memcpy on error?  The commit message does not describe why we want that, since gdrcopy is listed as an alternative, not in addition to.

wzamazon replied:
Removed the ifdefs in the code.

In the case gdrcopy is not available, we still want to use cudaMemcpy.

shefty commented on src/hmem_cuda.c:
```diff
@@ -118,8 +126,14 @@ int cuda_copy_to_dev(uint64_t device, void *dev, const void *host, size_t size)
+	err = cuda_gdrcopy_from_dev(device, host, dev, size);
+	if (!err)
+		return 0; 
+#endif
```
Similar comment as above.

wzamazon replied:
See my previous answer.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,292 @@
+	void *gdrcopy_user_ptr;
+
+	if (!gdrcopy_ops.gdr_copy_to_mapping)
+		return -FI_ENOSYS;
```
Can this just be an assert?

wzamazon replied:
Yes, it can and should be an assert, fixed in new version.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,292 @@
+	void *gdrcopy_user_ptr;
+
+	if (!gdrcopy_ops.gdr_copy_from_mapping)
+		return -FI_ENOSYS;
```
Can this be an assert?

wzamazon replied:
Yes, it can and should be an assert, fixed in new version.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,311 @@
+
+#endif /* ENABLE_CUDA_DLOPEN */
+
+gdr_t ofi_gdrcopy_open()
```
Could this be called during cuda_gdrcopy_hmem_init() have a static gdr_t handle be initialized?

wzamazon replied:
Sure, I will do that. We will need to add a lock to protect the gdr resource for multi thread environment.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,311 @@
+	return gdrcopy_ops.gdr_close(gdr);
+}
+
+ssize_t ofi_gdrcopy_reg(void *addr, size_t len, gdr_t gdr, struct ofi_gdrcopy_handle *gdrcopy)
```
I wonder if we need a new ofi HMEM operation to support this use case. For example, I have been looking at implementing an ofi_hmem_host_register(). Maybe we should have a ofi_hmem_dev_register() which returns an opaque handle to the caller.

wzamazon replied:
Sure, I will do that.

iziemba commented on src/hmem_cuda.c:
```diff
@@ -102,6 +102,12 @@ CUresult ofi_cuInit(unsigned int flags)
 {
+	int err;
+
+	err = cuda_gdrcopy_to_dev(device, dev, host, size);
```
Would it make sense to have the HMEM copy dispatch table call cuda_gdrcopy_to_dev() directly? Similar question for cuda_gdrcopy_from_dev().

wzamazon replied:
Good idea, will do that.

wzamazon commented:
@iziemba  I update the commit as you suggested Please take a look.

shefty commented:
If cudaMemcpy doesn't work as a general case, should gdrcopy be the default?  Why make this a configure option?

Given the problem it addresses, can we build this in always and allow switching based on an environment variable?

wzamazon commented:
The configure option is needed to provide the installation path of gdrcopy. Without it libfabric does not know where the header file is.

The problem with cudaMemcpy happens when libfabric is used by NCCL. My understanding is when libfabric is used by MPI, there was not such issue.

When configured with gdrcopy, gdrcopy is used by default. If user does not want to use gdrcopy, they configure libfabric without gdrcopy, so I am sure when would an environment variable will be useful.

shefty commented:
The issue is that there must be different builds of libfabric on a system, depending on whether NCCL or MPI is using it.  That's not ideal.  It would be better if a single build could handle both cases.

Why can't we search for gdrcopy the same as we do for cuda?  Letting the user specify a different install location is fine, but this doesn't seem like it should be an either-or build restriction with cudamemcpy

wzamazon commented:
@shefty Thanks. I see you point now.

> Why can't we search for gdrcopy the same as we do for cuda?

That is already happening with the --enable-gdrcopy-dlopen configure option. With that option, libfabric is not linked to gdrcopy library directly, but will use dlopen to search for gdrcopy library during runtime.

> Letting the user specify a different install location is fine, but this doesn't seem like it should be an either-or build restriction with cudamemcpy

--with-gdrcopy option and --with-cuda option are not either or relation. Instead, the --with-gdrcopy option is expected to be used together with --with-cuda. 

I will update the commit such that: when --with-gdrcopy is used, gdrcopy is used by default, user can use an environment variable to use cudaMemcpy.

wzamazon commented:
@shefty I added the environment variable FI_HMEM_CUDA_USE_GDRCOPY in new revision. By default it is 1.

shefty commented on include/ofi_hmem.h:
```diff
@@ -132,6 +141,18 @@ static inline int ofi_hmem_no_close_handle(void *ipc_ptr)
+int ofi_hmem_no_dev_deregister(uint64_t handle)
+{
+	return -FI_ENOSYS;
+}
```
If dev registration is not required, can this return success instead?

I'd like this change separated out (hold off for now), but mostly see if it needs to be aligned with PR #6176, which adds a host register call...

wzamazon replied:
Sure, return success here make sense.

I will move the abstraction of ```dev_register``` and ```dev_deregister``` to a separate PR then

wzamazon commented:
@shefty  In the new update, I removed ```dev_register``` and ```dev_deregister``` from ```hmem_ops``` and added two functions ```ofi_hmem_cuda_dev_register``` and ```ofi_hmem_cuda_dev_deregister```. Because this two are cuda specific operations.

iziemba commented on src/hmem_cuda.c:
```diff
@@ -132,6 +142,23 @@ int cuda_copy_from_dev(uint64_t device, void *host, const void *dev, size_t size
+		return cuda_gdrcopy_dev_register(mr_attr, handle);
+
+	*handle = mr_attr->device.cuda;
+	return 0;
```
FI_SUCCESS is used throughout this file instead of 0. Can it be used here as well?

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda.c:
```diff
@@ -53,6 +53,8 @@ struct cuda_ops {
 	CUresult (*cuInit)(unsigned int flags);
 };
 
+int hmem_cuda_use_gdrcopy;
```
static int hmem_cuda_use_gdrcopy;

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+};
+
+static gdr_t global_gdr;
+pthread_spinlock_t global_gdr_lock;
```
static pthread_spinlock_t global_gdr_lock;

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+{
+	int err;
+
+	cuda_gdrcopy_dl_hmem_init();
```
Check return value.

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+	global_gdr = global_gdrcopy_ops.gdr_open();
+	if (!global_gdr) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "gdr_open failed!\n");
+		return -FI_ENOMEM;
```
Does cuda_gdrcopy_dl_hmem_cleanup() need to be called on error?

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+		FI_WARN(&core_prov, FI_LOG_CORE, "close global_gdr_lock! err: %s\n",
+			strerror(err));
+	}
+
```
Missing cuda_gdrcopy_dl_hmem_cleanup()?

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+static void *gdrapi_handle;
+static struct gdrcopy_ops global_gdrcopy_ops;
+
+int cuda_gdrcopy_dl_hmem_init(void)
```
static

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+	return -FI_ENODATA;
+}
+
+int cuda_gdrcopy_dl_hmem_cleanup(void)
```
static

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+	.gdr_copy_from_mapping = gdr_copy_from_mapping
+};
+
+int cuda_gdrcopy_dl_hmem_init(void)
```
static

wzamazon replied:
Thanks! fixed in new update.

iziemba commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,367 @@
+	return FI_SUCCESS;
+}
+
+int cuda_gdrcopy_dl_hmem_cleanup(void)
```
static

wzamazon replied:
Thanks! fixed in new update.

wzamazon commented:
rebased on latest master.

shefty commented on include/ofi_hmem.h:
```diff
@@ -105,6 +105,15 @@ int cuda_hmem_cleanup(void);
 int cuda_host_register(void *ptr, size_t size);
 int cuda_host_unregister(void *ptr);
+int ofi_hmem_cuda_dev_register(struct fi_mr_attr *mr_attr, uint64_t *handle);
+int ofi_hmem_cuda_dev_deregister(uint64_t handle);
```
Rename to just cuda_dev_xxx, and drop the ofi_hmem_ prefix.  Though are these abstractions even needed?  I'm not finding where they are used.

wzamazon replied:
They are going to be used in https://github.com/ofiwg/libfabric/pull/6219. As utility function I think it is better to have the ofi prefix? like ofi_cudaMemcpy?

shefty replied:
We went with just the cuda prefix for the cuda calls, like the ones above this, to avoid prefix overload.

wzamazon replied:
Fixed in new revision.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+	gdrcopy->cuda_ptr = (void *)regbgn;
+	gdrcopy->length = reglen;
+
+	err = global_gdrcopy_ops.gdr_map(global_gdr, gdrcopy->mh, &gdrcopy->user_ptr, gdrcopy->length);
```
I'm only picking one spot, but please check line lengths throughout the patch and try to limit to roughly 80 characters.

wzamazon replied:
Fixed in new revision.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+
+int cuda_gdrcopy_hmem_cleanup(void)
+{
+	return -FI_ENOSYS;
```
nit: This can probably return success, since there's nothing to actually do.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+		return -FI_ENOMEM;
+
+	assert(global_gdr);
+	pthread_spin_lock(&global_gdr_lock);
```
Here and the unregister call, is a lock needed to serialize calling pin_buffer with map?  Is this the usage model gdr imposes on the user?

wzamazon replied:
The document of gdr does not mention about lock safety. From the source code, gdr_t is a wrapper for a file descriptor. gdr_pin_buffer will map gpu memory to that file descriptor, and gdr_map call mmap the file context to a user space pointer. So I believe lock is needed here.

shefty replied:
What is the lock protecting?  Why does it matter that pin_memory and map occur atomically wrt other callers?

wzamazon replied:
The lock is protecting ```global_gdr```, both pin_memory and gdr_map modify ```global_gdr``` so I put them inside lock.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+		assert(global_gdrcopy_ops.gdr_close);
+		global_gdrcopy_ops.gdr_close(global_gdr);
+		cuda_gdrcopy_dl_hmem_cleanup();
+		return err;
```
This is unlikely to fail anyway, but the return value should be -err.  For better maintenance, I would jump to a common exit point where cleanup can be unwound.

wzamazon replied:
fixed in new revision.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+	if (err) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "destroy global_gdr_lock failed! err: %s\n",
+			strerror(err));
+		ret = err;
```
Also unlikely to fail.  I would consider just dropping this check.  But if kept, the return value should be -err.

shefty replied:
reposting comment: Also unlikely to fail. I would consider just dropping this check. But if kept, the return value should be -err.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_to_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return 0;
```
function always returns success, so I would convert to void

wzamazon replied:
Converted to void in new revision.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_from_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return FI_SUCCESS;
```
function always returns success, so I would convert to void

wzamazon replied:
Done.

shefty commented on src/hmem_cuda_gdrcopy.c:
```diff
@@ -0,0 +1,380 @@
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	global_gdrcopy_ops.gdr_copy_from_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return FI_SUCCESS;
+}
```
The above functions could be combined into a single call, with a flag indicating the direction of the copy.

wzamazon replied:
Done.

shefty reviewed (commented):
See comments for a couple of minor changes.  I'm going to go ahead and merge this to make progress though.  Changes can come as a separate patch.
